Compiler failure when querying long timestamp type

```sql
CREATE TABLE memory.default.t(ts) AS VALUES (TIMESTAMP '2020-01-01 00:00:00.000000000');
SELECT * FROM memory.default.t WHERE ts = ts;
```

fails with 

```
Query 20200813_203153_00019_yq2rg failed: Compiler failed
io.prestosql.spi.PrestoException: Compiler failed
	at io.prestosql.sql.planner.LocalExecutionPlanner$Visitor.visitScanFilterAndProject(LocalExecutionPlanner.java:1325)
	at io.prestosql.sql.planner.LocalExecutionPlanner$Visitor.visitFilter(LocalExecutionPlanner.java:1185)
	at io.prestosql.sql.planner.LocalExecutionPlanner$Visitor.visitFilter(LocalExecutionPlanner.java:711)
	at io.prestosql.sql.planner.plan.FilterNode.accept(FilterNode.java:72)
	at io.prestosql.sql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:462)
	at io.prestosql.sql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:384)
	at io.prestosql.execution.SqlTaskExecutionFactory.create(SqlTaskExecutionFactory.java:75)
	at io.prestosql.execution.SqlTask.updateTask(SqlTask.java:390)
	at io.prestosql.execution.SqlTaskManager.updateTask(SqlTaskManager.java:383)
	at io.prestosql.server.TaskResource.createOrUpdateTask(TaskResource.java:132)
	at jdk.internal.reflect.GeneratedMethodAccessor354.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:76)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:148)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:191)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:200)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:103)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:493)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:415)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:104)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:277)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:272)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:268)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:316)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:298)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:268)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:289)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:256)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:703)
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:416)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:370)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:389)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:342)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:229)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:755)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1617)
	at io.airlift.http.server.TraceTokenFilter.doFilter(TraceTokenFilter.java:63)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at io.airlift.http.server.TimingFilter.doFilter(TimingFilter.java:51)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:545)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:717)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1300)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:485)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1215)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
	at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:173)
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:59)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:500)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:547)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.google.common.util.concurrent.UncheckedExecutionException: java.lang.RuntimeException: java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    io/prestosql/$gen/CursorProcessor_20200813_203153_48.filter(Lio/prestosql/spi/connector/ConnectorSession;Lio/prestosql/spi/connector/RecordCursor;)Z @67: invokedynamic
  Reason:
    Type 'java/lang/Object' (current frame, stack[1]) is not assignable to 'io/prestosql/spi/type/LongTimestamp'
  Current Frame:
    bci: @67
    flags: { }
    locals: { 'io/prestosql/$gen/CursorProcessor_20200813_203153_48', 'io/prestosql/spi/connector/ConnectorSession', 'io/prestosql/spi/connector/RecordCursor', integer }
    stack: { 'java/lang/Object', 'java/lang/Object' }
  Bytecode:
    0000000: 033e 2c03 b900 4102 0099 0009 043e 01a7
    0000010: 000a 2c03 b900 4502 001d 9900 0857 03a7
    0000020: 0037 2c03 b900 4102 0099 0009 043e 01a7
    0000030: 000a 2c03 b900 4502 001d 9900 0957 5703
    0000040: a700 16ba 0052 0000 59c7 000a 043e 5703
    0000050: a700 06b6 0057 1d99 0005 5703 ac
  Stackmap Table:
    append_frame(@18,Integer)
    same_locals_1_stack_item_frame(@25,Object[#4])
    same_locals_1_stack_item_frame(@34,Object[#4])
    same_locals_1_stack_item_frame(@50,Object[#4])
    full_frame(@57,{Object[#2],Object[#90],Object[#23],Integer},{Object[#4],Object[#4]})
    full_frame(@67,{Object[#2],Object[#90],Object[#23],Integer},{Object[#4],Object[#4]})
    same_locals_1_stack_item_frame(@83,Object[#84])
    same_locals_1_stack_item_frame(@86,Integer)
    same_locals_1_stack_item_frame(@92,Integer)

	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2051)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3951)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3974)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4958)
	at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4964)
	at io.prestosql.sql.gen.ExpressionCompiler.compileCursorProcessor(ExpressionCompiler.java:83)
	at io.prestosql.sql.planner.LocalExecutionPlanner$Visitor.visitScanFilterAndProject(LocalExecutionPlanner.java:1288)
	... 71 more
Caused by: java.lang.RuntimeException: java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    io/prestosql/$gen/CursorProcessor_20200813_203153_48.filter(Lio/prestosql/spi/connector/ConnectorSession;Lio/prestosql/spi/connector/RecordCursor;)Z @67: invokedynamic
  Reason:
    Type 'java/lang/Object' (current frame, stack[1]) is not assignable to 'io/prestosql/spi/type/LongTimestamp'
  Current Frame:
    bci: @67
    flags: { }
    locals: { 'io/prestosql/$gen/CursorProcessor_20200813_203153_48', 'io/prestosql/spi/connector/ConnectorSession', 'io/prestosql/spi/connector/RecordCursor', integer }
    stack: { 'java/lang/Object', 'java/lang/Object' }
  Bytecode:
    0000000: 033e 2c03 b900 4102 0099 0009 043e 01a7
    0000010: 000a 2c03 b900 4502 001d 9900 0857 03a7
    0000020: 0037 2c03 b900 4102 0099 0009 043e 01a7
    0000030: 000a 2c03 b900 4502 001d 9900 0957 5703
    0000040: a700 16ba 0052 0000 59c7 000a 043e 5703
    0000050: a700 06b6 0057 1d99 0005 5703 ac
  Stackmap Table:
    append_frame(@18,Integer)
    same_locals_1_stack_item_frame(@25,Object[#4])
    same_locals_1_stack_item_frame(@34,Object[#4])
    same_locals_1_stack_item_frame(@50,Object[#4])
    full_frame(@57,{Object[#2],Object[#90],Object[#23],Integer},{Object[#4],Object[#4]})
    full_frame(@67,{Object[#2],Object[#90],Object[#23],Integer},{Object[#4],Object[#4]})
    same_locals_1_stack_item_frame(@83,Object[#84])
    same_locals_1_stack_item_frame(@86,Integer)
    same_locals_1_stack_item_frame(@92,Integer)

	at io.airlift.bytecode.ClassGenerator.defineClasses(ClassGenerator.java:181)
	at io.airlift.bytecode.ClassGenerator.defineClass(ClassGenerator.java:117)
	at io.prestosql.util.CompilerUtils.defineClass(CompilerUtils.java:63)
	at io.prestosql.util.CompilerUtils.defineClass(CompilerUtils.java:57)
	at io.prestosql.sql.gen.ExpressionCompiler.compileProcessor(ExpressionCompiler.java:167)
	at io.prestosql.sql.gen.ExpressionCompiler.compile(ExpressionCompiler.java:134)
	at io.prestosql.sql.gen.ExpressionCompiler.lambda$new$0(ExpressionCompiler.java:70)
	at com.google.common.cache.CacheLoader$FunctionToCacheLoader.load(CacheLoader.java:165)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045)
	... 77 more
Caused by: java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    io/prestosql/$gen/CursorProcessor_20200813_203153_48.filter(Lio/prestosql/spi/connector/ConnectorSession;Lio/prestosql/spi/connector/RecordCursor;)Z @67: invokedynamic
  Reason:
    Type 'java/lang/Object' (current frame, stack[1]) is not assignable to 'io/prestosql/spi/type/LongTimestamp'
  Current Frame:
    bci: @67
    flags: { }
    locals: { 'io/prestosql/$gen/CursorProcessor_20200813_203153_48', 'io/prestosql/spi/connector/ConnectorSession', 'io/prestosql/spi/connector/RecordCursor', integer }
    stack: { 'java/lang/Object', 'java/lang/Object' }
  Bytecode:
    0000000: 033e 2c03 b900 4102 0099 0009 043e 01a7
    0000010: 000a 2c03 b900 4502 001d 9900 0857 03a7
    0000020: 0037 2c03 b900 4102 0099 0009 043e 01a7
    0000030: 000a 2c03 b900 4502 001d 9900 0957 5703
    0000040: a700 16ba 0052 0000 59c7 000a 043e 5703
    0000050: a700 06b6 0057 1d99 0005 5703 ac
  Stackmap Table:
    append_frame(@18,Integer)
    same_locals_1_stack_item_frame(@25,Object[#4])
    same_locals_1_stack_item_frame(@34,Object[#4])
    same_locals_1_stack_item_frame(@50,Object[#4])
    full_frame(@57,{Object[#2],Object[#90],Object[#23],Integer},{Object[#4],Object[#4]})
    full_frame(@67,{Object[#2],Object[#90],Object[#23],Integer},{Object[#4],Object[#4]})
    same_locals_1_stack_item_frame(@83,Object[#84])
    same_locals_1_stack_item_frame(@86,Integer)
    same_locals_1_stack_item_frame(@92,Integer)

	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:398)
	at com.google.common.reflect.Reflection.initialize(Reflection.java:65)
	at io.airlift.bytecode.ClassGenerator.defineClasses(ClassGenerator.java:177)
	... 88 more
```

This is  likely caused by a missing cast in the generated code that reads from RecordCursor.

--- a/presto-main/src/main/java/io/prestosql/sql/gen/CursorProcessorCompiler.java
+++ b/presto-main/src/main/java/io/prestosql/sql/gen/CursorProcessorCompiler.java
@@ -15,7 +15,6 @@ package io.prestosql.sql.gen;
 
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.ImmutableSet;
-import com.google.common.primitives.Primitives;
 import io.airlift.bytecode.BytecodeBlock;
 import io.airlift.bytecode.BytecodeNode;
 import io.airlift.bytecode.ClassDefinition;
@@ -305,9 +304,6 @@ public class CursorProcessorCompiler
                 Variable wasNullVariable = scope.getVariable("wasNull");
 
                 Class<?> javaType = type.getJavaType();
-                if (!javaType.isPrimitive() && javaType != Slice.class) {
-                    javaType = Object.class;
-                }
 
                 IfStatement ifStatement = new IfStatement();
                 ifStatement.condition()
@@ -322,8 +318,23 @@ public class CursorProcessorCompiler
 
                 ifStatement.ifFalse()
                         .getVariable(cursorVariable)
-                        .push(field)
-                        .invokeInterface(RecordCursor.class, "get" + Primitives.wrap(javaType).getSimpleName(), javaType, int.class);
+                        .push(field);
+                if (javaType == boolean.class) {
+                    ifStatement.ifFalse().invokeInterface(RecordCursor.class, "getBoolean", boolean.class, int.class);
+                }
+                else if (javaType == long.class) {
+                    ifStatement.ifFalse().invokeInterface(RecordCursor.class, "getLong", long.class, int.class);
+                }
+                else if (javaType == double.class) {
+                    ifStatement.ifFalse().invokeInterface(RecordCursor.class, "getDouble", double.class, int.class);
+                }
+                else if (javaType == Slice.class) {
+                    ifStatement.ifFalse().invokeInterface(RecordCursor.class, "getSlice", Slice.class, int.class);
+                }
+                else {
+                    ifStatement.ifFalse().invokeInterface(RecordCursor.class, "getObject", Object.class, int.class)
+                            .checkCast(javaType);
+                }
 
                 return ifStatement;
             }

